### PR TITLE
Split void into two types

### LINKED ISSUE #6077: Rename the void value currently named "_void"
There is currently a "value" named `_void` that has type `void`.  We can't name both the value and the type the same thing, and don't want to continue using the leading underscore.

This value will probably be most frequently used in variable or field declarations, but can be used in some other places as well:

```
var myRemovableVar: void = _void;
var debugString = if debug then "print me" else _void;

writeln(_void); // prints a blank line
```

Brainstorming a few possibilities:
* `empty`
* `nothing`
* `unused`
* `none` // I believe python uses `none` for a similar idea

I'm happy to take other suggestions, as none of those seem great to me.

### LINKED ISSUE #12885: Naming for the two void concepts and the void value
Issue #12208 seems to have come mostly to a consensus about the idea of separating the two concepts that we currently call `void`.  There was one dissenter, but the level of dissent was fairly low.  This new issue is to discuss naming the three user-visible things.  There are two types and a value.

I propose that we take @bradcray's second choice from that issue:

```chapel
proc foo(): nothing { ...return;...}  // I return no value
proc bar(): void { ...return voidValue;...}  // I return an empty value
```

I was initially feeling that we should borrow from Python's `none` and `noneType` (capitalization changes intentional), but this could lead to confusion due to possibly subtle differences between the concepts in Chapel vs. Python.

The idea in his first choice seems too easy to overlook to me.  I think a person making the effort to declare the return type is going to want it to stand out more than a single character.

```chapel
proc foo(): where abc { ...return;...}  // I return no value
proc bar()  where abc { ...return 42;...}  // I return a value
```

### PR DESCRIPTION
**Pull request by daviditen**
Split `void` into two distinct types. Now `void` is the type of a function
that doesn't return anything, e.g.

```chapel
proc foo(): void {
  writeln("Hello");
  return;
}
```

The second type is for what has previously been called the `void value`.
This type is now called `nothing` and has a single value of `none`.  It is
the value and type that will cause variables and fields to be removed by
the compiler.

```chapel
record R {
  param useRng = false;
  // The 'rng' field is removed if 'useRng' is false
  var rng = if useRng then 1..100 else none;
}
```

Updated the modules and tests to use the right types based on context.

Closes #12885 and closes #6077.

### CODE REVIEW

**Comment by daviditen** [comment]
Passed Linux64 paratest.

**Comment by mppf** [comment]
>   var rng = if useRng then 1..100 else none;

Wow that's not a good RNG at all

**Comment by mppf** [comment]
I was expecting this to update doc/rst/technotes/voidVariables.rst  but it doesn't appear to.

**Comment by mppf** [comment]
You should have the PR link to (and probably close) #12885

File: compiler/resolution/functionResolution.cpp [hunk 1]
```diff
@@ -5275,7 +5275,11 @@ static Type* moveDetermineRhsType(CallExpr* call) {
   if (retval == dtVoid) {
     if (CallExpr* rhsCall = toCallExpr(rhs)) {
       if (FnSymbol* rhsFn = rhsCall->resolvedFunction()) {
-        if (rhsFn->hasFlag(FLAG_VOID_NO_RETURN_VALUE) == true) {
+        // `expandExternArrayCalls` can add void assignments when
+        // the extern function has an explicit void return type.
```
Comment (mppf) [inline compiler/resolution/functionResolution.cpp]: Is there a way to handle it by adjusting expandExternArrayCalls?

**Comment by daviditen** [comment]
I've left this as a purely functional update so far.  doc/rst/technotes/voidVariables.rst and test/release/examples/primers/voidVariables.chpl both need updates.

File: modules/standard/Types.chpl [hunk 2]
```diff
@@ -60,7 +60,7 @@ as defined by the language specification.
 */
 pragma "no instantiation limit"
 proc isPrimitiveType(type t) param return
-  isVoidType(t) || isBoolType(t) || isIntType(t) || isUintType(t) ||
+  isNothingType(t) || isBoolType(t) || isIntType(t) || isUintType(t) ||
```
Comment (mppf) [inline modules/standard/Types.chpl]: I think `void` should still be a primitive type

File: compiler/resolution/functionResolution.cpp (hunk 1 above)
Comment (daviditen) [reply compiler/resolution/functionResolution.cpp]: Not fully.  If the extern function doesn't specify a return type it is handled correctly there. If it does specify a return type then the type is not known at that point.

```
extern proc foo(a: [] real); // this is handled now
extern proc bar(a: [] real): void; // this void is not resolved yet in expandExternArrayCalls
type t = void;
extern proc baz(a: [] real): t; // this void would be very difficult to resolve at that point
```

**Comment by daviditen** [comment]
Thanks @mppf! I've made your requested changes and am going to run another round of regular testing once my CHPL_LLVM=llvm + --llvm testing completes soon.

**Comment by mppf** [comment]
Looks like Travis is complaining about uses of `none` in `BLAS.chpl`

**Comment by daviditen** [comment]
> Looks like Travis is complaining about uses of none in BLAS.chpl

Yep, I'm looking into that.

**Comment by daviditen** [comment]
There were 8 test failures in --llvm testing :( :
[Error matching compiler output for arrays/deitz/jacobi-unoptimized]
[Error matching debug info for llvmDebug_test]
[Error 1 running sub_test in test/llvm/llvmDebug]
[Error matching compiler output for npb/cg/bradc/cg-sparse-partred]
[Error matching compiler output for reductions/partial/DR-with-shaped-exprs]
[Error matching compiler output for reductions/partial/DRtest]
[Error matching compiler output for trivial/shannon/condReturn]
[Error matching compiler output for types/void/void-assign]

**Comment by daviditen** [comment]
All of the --llvm errors are of the forms:
```
error: Could not find C type for nothing
error: Could not find C type internal_void - perhaps it is a complex macro?
```

I'm guessing that they have an easy fix.

**Comment by mppf** [comment]
@daviditen - there is a strcmp with "void" in compiler/codegen/symbol.cpp codegenFunctionType around line 1678. I recommend that you set the cname of nothing and internal_void to `void`. However it's also reasonable to update that code to check if retType == dtNothing or dtVoid.

**Comment by daviditen** [comment]
@mppf Thanks, that cleared out most of them.

These two still fail in the same way, I'll dig into them more to see what's going on.
[Error matching compiler output for arrays/deitz/jacobi-unoptimized]
[Error matching compiler output for types/void/void-assign]

**Comment by daviditen** [comment]
Testing is now clean for both --llvm and regular paratest other than these tests which have been fixed since my last merge with master:

```
extern/ferguson/externblock/readme_example-error.chpl
classes/delete-free/isClassTest.chpl
```

@mppf, is it OK to merge?

**Comment by mppf** [comment]
@daviditen - yes, LGTM

**Comment by daviditen** [comment]
Thank you!

**Pull request merged**